### PR TITLE
[13.0][ADD] delivery_price_rule_volumetric_weight: Add Volumetric Weight to delivery price rules

### PR DESCRIPTION
**Pull request by victoralmau**
Add Volumetric Weight to delivery price rules.

Locked by:
- [x] `product_dimension_volumetric_weight` https://github.com/OCA/product-attribute/pull/889

Please @pedrobaeza and @chienandalu can you review it?

@Tecnativa TT30670

### CODE REVIEW

File: delivery_price_rule_volumetric_weight/models/delivery_carrier.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    def _get_price_available(self, order):
+        self.ensure_one()
+        self = self.sudo()
+        order = order.sudo()
+        total = weight = volumetric_weight = volume = quantity = 0
+        total_delivery = 0.0
+        for line in order.order_line:
+            if line.state == "cancel":
+                continue
```
Comment (chienandalu) [inline delivery_price_rule_volumetric_weight/models/delivery_carrier.py]: Why don't you simply filter these lines?

File: delivery_price_rule_volumetric_weight/models/delivery_carrier.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            if line.is_delivery:
+                total_delivery += line.price_total
+            if not line.product_id or line.is_delivery:
+                continue
```
Comment (chienandalu) [inline delivery_price_rule_volumetric_weight/models/delivery_carrier.py]: Same here

**Review by chienandalu — COMMENTED** [review]
Tested :+1: Minor code comment:

File: delivery_price_rule_volumetric_weight/models/delivery_carrier.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        order = order.sudo()
+        total = weight = volumetric_weight = volume = quantity = 0
+        total_delivery = 0.0
+        for line in order.order_line.filtered(
+            lambda x: x.state != "cancel" and x.product_id
+        ):
+            if line.is_delivery:
+                total_delivery += line.price_total
+            if line.is_delivery:
+                continue
```
Comment (chienandalu) [inline delivery_price_rule_volumetric_weight/models/delivery_carrier.py]: No need to check twice :slightly_smiling_face: 
```suggestion
            if line.is_delivery:
                total_delivery += line.price_total
                continue
```

File: delivery_price_rule_volumetric_weight/models/delivery_carrier.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            quantity += qty
+        total = (order.amount_total or 0.0) - total_delivery
+        total = order.currency_id._convert(
+            total,
+            order.company_id.currency_id,
+            order.company_id,
+            order.date_order or fields.Date.today(),
+        )
+        _self = self.with_context(volumetric_weight=volumetric_weight)
+        return _self._get_price_from_picking(total, weight, volume, quantity)
```
Comment (pedrobaeza) [inline delivery_price_rule_volumetric_weight/models/delivery_carrier.py]: Do this is a more compatible way:
```suggestion
        for line in order.order_line.filtered(
            lambda x: x.state != "cancel" and x.product_id and not x.is_delivery
        ):
            volumetric_weight += (line.product_id.volumetric_weight or 0.0) * qty
        _self = self.with_context(volumetric_weight=volumetric_weight)
        return super(DeliveryCarrier, _self)._get_price_available(order)
```

File: delivery_price_rule_volumetric_weight/models/delivery_carrier.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+    def _get_price_from_picking(self, total, weight, volume, quantity):
+        price = 0.0
+        criteria_found = False
+        price_dict = {
```
Comment (pedrobaeza) [inline delivery_price_rule_volumetric_weight/models/delivery_carrier.py]: You can try to propose to Odoo a hook that return the `price_dict` dictionary for not having to override whole method.

File: delivery_price_rule_volumetric_weight/README.rst [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+.. |badge4| image:: https://img.shields.io/badge/weblate-Translate%20me-F47D42.png
+    :target: https://translation.odoo-community.org/projects/delivery-carrier-13-0/delivery-carrier-13-0-delivery_price_rule_volumetric_weight
+    :alt: Translate me on Weblate
+.. |badge5| image:: https://img.shields.io/badge/runbot-Try%20me-875A7B.png
+    :target: https://runbot.odoo-community.org/runbot/99/13.0
+    :alt: Try me on Runbot
+
+|badge1| |badge2| |badge3| |badge4| |badge5| 
+
+These addon add the possibility to choose volumetric weight in delivery price rules.
```
Comment (pedrobaeza) [inline delivery_price_rule_volumetric_weight/README.rst]:
```suggestion
This module adds the possibility of choosing volumetric weight in delivery price rules.
```

File: delivery_price_rule_volumetric_weight/README.rst [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+Usage
+=====
+
+To use this module :
+
+#. Go to Sale > Orders > Quotations.
+#. Create new order and set some product with volumetric weight.
+#. Set some carrier with volumetric weight in rules.
+#. Delivery price have changed according carrier rules.
```
Comment (pedrobaeza) [inline delivery_price_rule_volumetric_weight/README.rst]:
```suggestion
#. Delivery price has changed according carrier rules.
```

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: delivery_price_rule_volumetric_weight/models/delivery_carrier.py (hunk 5 above)
Comment (victoralmau) [reply delivery_price_rule_volumetric_weight/models/delivery_carrier.py]: https://github.com/odoo/odoo/pull/74529

**Review by pedrobaeza — APPROVED** [review]

**Comment by pedrobaeza** [comment]
I have pushed OCB for having the hook, so let's merge it:

/ocabot merge nobump

**Pull request merged**